Libraries/LibCrypto/NumberTheory/ModularFunctions.h

```diff
@@ -0,0 +1,202 @@
+                if (d.words()[0] % 2 == 1) {
+                    d = d.plus(b);
+                }
+                u = u.divided_by(2).quotient;
+                d = d.divided_by(2).quotient;
+            }
+        }
+        v = v.minus(u);
+        x = x.plus(d);
+        while (v.words()[0] % 2 == 0) {
+            if (x.words()[0] % 2 == 1) {
+                x = x.plus(b);
+            }
+            v = v.divided_by(2).quotient;
+            x = x.divided_by(2).quotient;
+        }
+    }
+    return x.divided_by(b).remainder;
+}
+
+static auto ModularPower(const UnsignedBigInteger& b, const UnsignedBigInteger& e, const UnsignedBigInteger& m) -> UnsignedBigInteger
+{
+    if (m == 1)
+        return 0;
+
+    UnsignedBigInteger ep { e };
+    UnsignedBigInteger base { b };
+    UnsignedBigInteger exp { 1 };
+
+    while (!(ep < 1)) {
+#ifdef NT_DEBUG
+        dbg() << ep.to_base10();
+#endif
+        if (ep.words()[0] % 2 == 1) {
+            exp = exp.multiplied_by(base).divided_by(m).remainder;
+        }
+        ep = ep.divided_by(2).quotient;
+        base = base.multiplied_by(base).divided_by(m).remainder;
+    }
+    return exp;
+}
+
+static auto GCD(const UnsignedBigInteger& a, const UnsignedBigInteger& b) -> UnsignedBigInteger
+{
+    UnsignedBigInteger a_ { a }, b_ { b };
+    for (;;) {
+        if (a_ == 0)
+            return b_;
+        b_ = b_.divided_by(a_).remainder;
+        if (b_ == 0)
+            return a_;
+        a_ = a_.divided_by(b_).remainder;
+    }
+}
+
+static auto LCM(const UnsignedBigInteger& a, const UnsignedBigInteger& b) -> UnsignedBigInteger
+{
+    auto temp = GCD(a, b);
+
+    auto div = a.divided_by(temp);
+
+#ifdef NT_DEBUG
+    dbg() << "quot: " << div.quotient << " rem: " << div.remainder;
+#endif
+    return temp == 0 ? 0 : (a.divided_by(temp).quotient.multiplied_by(b));
+}
+
+template<size_t test_count>
+static bool MR_primality_test(UnsignedBigInteger n, const Vector<UnsignedBigInteger, test_count>& tests)
+{
+    auto prev = n.minus({ 1 });
+    auto b = prev;
+    auto r = 0;
+
+    auto div_result = b.divided_by(2);
+    while (div_result.quotient == 0) {
+        div_result = b.divided_by(2);
+        b = div_result.quotient;
+        ++r;
+    }
+
+    for (size_t i = 0; i < tests.size(); ++i) {
+        auto return_ = true;
+        if (n < tests[i])
+            continue;
+        auto x = ModularPower(tests[i], b, n);
+        if (x == 1 || x == prev)
+            continue;
+        for (auto d = r - 1; d != 0; --d) {
+            x = ModularPower(x, 2, n);
+            if (x == 1)
+                return false;
+            if (x == prev) {
+                return_ = false;
+                break;
+            }
+        }
+        if (return_)
+            return false;
+    }
+
+    return true;
+}
+
+static UnsignedBigInteger random_number(const UnsignedBigInteger& min, const UnsignedBigInteger& max)
+{
+    auto range = max.minus(min);
```

Comment:
Maybe add an `ASSERT(min < max)` ? Otherwise, this would result in `range` being an invalid `UnsignedBigInteger` but its trimmed length doesn't take that into account.

Reply:
:thinking: what would it do in that case, I wonder?
I'll add the assert~